### PR TITLE
saferhtml

### PR DESCRIPTION
**Pull request by saggi-dw**
Attempt to make the use of HTML a bit safer
- deny list for tags
- allow-list for <del>tags</del> attributes
Use the simple_html_dom.php for this purpose

### CODE REVIEW

**Comment by saggi-dw** [comment]
Hi @splitbrain 
I know you don't have much time to deal with it. Would this change make the concept better? I would appreciate an opinion....

File: syntax/htmlphpok.php [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        }
+        return true;
+    }
+
+    public function RemoveBadHTML($match, $addTag = false)
+    {
+        $htmlreturn = '';
+        $deniedTags = self::GetDeniedTags();
+        $allowedAttributes = self::GetAllowedAttributes();
+        if(!count($allowedAttributes) && !count($deniedTags)) {
```
Comment (fiwswe) [inline syntax/htmlphpok.php]: If I am reading this right then:
  * A corner case: Why aren't tags without any allowed attributes acceptable? Example: `<b>bold</b>`
  * Another one: Why can't attributes be filtered for all tags (no denied tags) ?

A style nit: Instead of `!count(…)` why not use `empty(…)`?

File: lang/en/settings.php [hunk 2]
```diff
@@ -8,4 +8,6 @@
 // keys need to match the config setting name
 $lang['htmlok']      = 'Allow embedded HTML. Do not enable if you cannot trust the editors of your wiki content!';
 $lang['phpok']       = 'Allow embedded PHP. Do not enable if you cannot trust the editors of your wiki content!';
-
+$lang['deniedtags']  = 'The Plugin uses an "deny list" of not interpreted HTML Tags. Default: [comment,embed,link,listing,meta,noscript,object,plaintext,script,xmp]';
+$lang['showdenied']  = 'Show content for removed tags as preformatted text (shows also a hint about removed tag)';
+$lang['allowedattr'] = 'The Plugin uses an "allow-list" approach and accepts only the following attributes:';
```
Comment (fiwswe) [inline lang/en/settings.php]: Nit: Missing newline at the end of the file.

File: syntax/htmlok.php [hunk 3]
```diff
@@ -22,7 +22,7 @@ public function getPType()
     /** @inheritDoc */
     public function getSort()
     {
-        return 190;
+        return 189;
```
Comment (fiwswe) [inline syntax/htmlok.php]: While I agree with this change (makes the plugin override the default in pre "Jack Jackrum" DokuWiki), shouldn't this be a separate PR?

File: syntax/phpblock.php [hunk 4]
```diff
@@ -22,7 +22,7 @@ public function getPType()
     /** @inheritDoc */
     public function getSort()
     {
-        return 190;
+        return 180;
```
Comment (fiwswe) [inline syntax/phpblock.php]: While I agree with this change (makes the plugin override the default in pre "Jack Jackrum" DokuWiki), shouldn't this be a separate PR?

Also why 180 and not 189 as for the generic case (`syntax_plugin_htmlok_htmlphpok`)? As a matter of fact, why override the generic case at all?

File: syntax/phpok.php [hunk 5]
```diff
@@ -22,7 +22,7 @@ public function getPType()
     /** @inheritDoc */
     public function getSort()
     {
-        return 190;
+        return 179;
```
Comment (fiwswe) [inline syntax/phpok.php]: While I agree with this change (makes the plugin override the default in pre "Jack Jackrum" DokuWiki), shouldn't this be a separate PR?

Also why 179 and not 189 as for the generic case (`syntax_plugin_htmlok_htmlphpok`)? As a matter of fact, why override the generic case at all?

**Review by fiwswe — COMMENTED** [review]
I Like the basic idea :-)

I'm not sure building a complete XML/HTML parser from scratch, even if you you didn't need to write the code yourself, is the best way to go. But as a first step it will work, I guess. And it avoids dependencies on PHP xml extensions.

**Comment by saggi-dw** [comment]
> I Like the basic idea :-)
> 
> I'm not sure building a complete XML/HTML parser from scratch, even if you you didn't need to write the code yourself, is the best way to go. But as a first step it will work, I guess. And it avoids dependencies on PHP xml extensions.

Thank you for the review. I have made some changes, however I can't get it into this branch. 
I now have the list of tags as an "allow list", which seems to fit better.
I agree with you that the customization of the sorting would be better in a separate commit.

File: syntax/htmlphpok.php (hunk 1 above)
Comment (saggi-dw) [reply syntax/htmlphpok.php]: > A style nit: Instead of `!count(…)` why not use `empty(…)`?

Because I can't always think of everything that makes sense ...
Thanks for the hint

**Comment by splitbrain** [comment]
I haven't looked into this too deep. In general: using a HTML sanitizer of course makes the whole concept safer (if not 100% safe). I would suggest using a well established library for that (not sure if you already do).

However for the usecase of "restoring" the old functionality this might already be too much. People using HTML to embed custom JavaScript won't want this feature.

**Comment by saggi-dw** [comment]
Thanks to both of you.

> In general: using a HTML sanitizer of course makes the whole concept safer (if not 100% safe). I would suggest using a well established library for that (not sure if you already do).

I looked at http://htmlpurifier.org/ and realized that this is a huge monster. So it's a pretty simple solution that should make it just a little bit safer. 

> However for the usecase of "restoring" the old functionality this might already be too much. People using HTML to embed custom JavaScript won't want this feature.

I think that's exactly the point. I'll finish this (with the helpful hints) and leave the PR open. Maybe the time will come that it will be needed.

**Comment by Klap-in** [comment]
This functionality feels partly as redoing https://www.dokuwiki.org/plugin:adhoctags
If each plugin does its main job good, it is in my opinion, for more users clear why they need each plugin. If you would recommend that people should consider the adhoctags plugin instead of htmlok, maybe an recommendation on the wiki page help the people already a lot (and is of course less work for maintaining this plugin ;-) )

**Comment by saggi-dw** [comment]
> maybe an recommendation on the wiki page help the people already a lot (and is of course less work for maintaining this plugin ;-) )

You're absolutely right, I think I was just looking for a way to exercise....
A recommendation in the wiki to better alternatives is certainly the best choice!

**Comment by saggi-dw** [comment]
@Klap-in [Alternatives listed](https://www.dokuwiki.org/faq:html#is_html_really_needed) and a [note on the plugin page](https://www.dokuwiki.org/plugin:htmlok). Would this be the implementation of your idea? I thought at the FAQ it is well kept, because several plugins provide the function.

**Comment by Klap-in** [comment]
Well explained.
